packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -48,7 +48,37 @@ export 'package:test_api/test_api.dart' hide
 /// Signature for callback to [testWidgets] and [benchmarkWidgets].
 typedef WidgetTesterCallback = Future<void> Function(WidgetTester widgetTester);
 
-/// Runs the [callback] inside the Flutter test environment.
+/// Runs the setup [callback] inside the Flutter test environment.
+///
+/// This is useful if you would like to run [setUp] code that needs to execute
+/// in the same async zone as the test code.
+///
+/// The following test code would never complete without using [setUpWidgets]
+/// because the `foo` Future would be created outside of the FakeAsync zone that
+/// [testWidgets] sets up. By using [setUpWidgets], we ensure the setup code
+/// runs in the same [FakeAsync] zone which allows us to advance the clock and
+/// let the Future resolve during [testWidgets].
+///
+/// ```dart
+/// void main() {
+///   Future<int> five;
+///   setUpWidgets((WidgetTester tester) async {
+///     five = Future.delayed(const Duration(seconds: 5), () => 5);
+///     await tester.pump(const Duration(seconds: 5));
+///   });
+///
+///   testWidgets('Hello world smoke test', (WidgetTester tester) async {
+///     expect(await five, 5);
+///   });
+/// }
+/// ```
+void setUpWidgets(WidgetTesterCallback callback) {
```

Comment:
For symmetry wouldn't we also need a tearDownWidgets method?

Reply:
Yes. @chunhtai and I talked about this.

Unfortunately, teardown is a lot more tricky to achieve with the current design of `testWidgets()`. It registers its own cleanup function right before running the test which runs as the *first* teardown (since they are run in reverse). This cleanup destroys the fake async zone.

I tried a couple of approaches but did not find one that I liked so far.

Reply:
I start to feel maybe we should stop leveraging the setUp/teardown method in test library, maybe we should just store all the setupwidgets and teardownwidgets in a queue in the binding and just run those as part of  runTest

Reply:
That's one of the options I tried. The tricky bit is that tests have _scoping_, via groups. Consider:

```dart
foo(int i) => print('foo $i!');

void main() {
  setUp(() => print('Outer setup'));
  foo(0);
  group('Outer group', () {
    foo(1);
    setUp(() => print('Inner setup'));
    test('test1', () => print('test1'));
  });
  test('test2', () => print('test2'));
}
```

This will print:

```
foo 0!
foo 1!
Outer setup
Inner setup
test1
Outer setup
test2
```

If we were to just add callbacks, they will be executed for every test regardless of their group scope. I *really* don't want to be in the business of implementing test_api scope :-). That's why we delegate to package:test APIs.

Reply:
OK so there are three ways we can go about this:

#### Don't do tearDownWidgets().
Symmetry aside, the invariants of the widget tester dictate that there should not be any unresolved timers/futures the fake async zone after test completes. I don't believe tearDownWidgets is going to be useful by design. Users can simply use normal tearDown. We can still expose the API for completeness but delegate to normal tearDown instead of test binding.

Now assuming that we still have a use case for running teardown in fake async:

#### Add a new API to package:test_api
This would allow adding teardown callbacks to head of the queue. Like `addFinalTeardown()` or some such.

#### Lazily eliminate fake async
We could keep a counter for unfinished teardowns and decrement that counter every time a teardown is done. Then we can do something like this in post test:

```dart
outstandingTeardowns.onZero.listen((_) => _fakeAsync = null);
```

Right now, I am leaning towards first option and documenting it. I really would like to avoid complicating this code further without a real use case. Let me know what you think.

Reply:
> I start to feel maybe we should stop leveraging the setUp/teardown method in test library, maybe we should just store all the setupwidgets and teardownwidgets in a queue in the binding and just run those as part of runTest

This is what I did for [flutter_test_ui](https://pub.dev/packages/flutter_test_ui), which works fine in practise, but is a bit hacky to be honest.

Reply:
Thanks! That's actually a fourth way. You were rather clever to use `setUp` function for teardowns to sidestep the scoping problem I mention. It is still too complex to be added without a real use case.

Therefore, I still vote for (1). We add a tearDownWidgets() function that simply does this:

```dart
class TeardownWidgetTester implements WidgetTester {
  @override
  dynamic noSuchMethod(Invocation invocation) {
    throw UnsupportedError('Explain here why WidgetTester cannot be used in teardown and encourage people to put their use case to a predefined github issue.');
  }
}

/// Runs the teardown [callback] in the test zone.
///
/// Note that the callback does *not* run in the test binding zone and therefore [WidgetTester] cannot be used.
/// If you have a use case for it, please update issue <link> and explain what your use case is.
void tearDownWidgets(WidgetTesterCallback callback) {
  tearDown(() async => await callback(TeardownWidgetTester()));
}
```

Reply:
Opened https://github.com/flutter/flutter/issues/58172 and pushed a new version. Please take a look.

Reply:
We only recently started using flutter_test_ui and this is of course only our project, so for what it is worth: `testUI` is used 266 times in our code base, `setUpUI` 201 times, and `tearDownUI` only once (to let timers expire that some tests aren't concerned with).

Reply:
This discussion has been very illuminating. Thanks @spkersten.

Your example in the homepage also gave me a test case which made me realize that my approach does not work.

Testing framework not only sets up and destroys zone in testBody, it also destroys any widget trees prior to executing the test. 

The only way to make this work without significantly changing the package:flutter_test semantics is do exactly what you have done :-). I will simply make it part of the testing framework.